### PR TITLE
perf improvements from buffer reuse

### PR DESCRIPTION
**Pull request by alisterburt**
Hit some perf issues with this - mainly due to unnecessary tensor creation

Will likely cleanup before merging but want to test perf properly on gpu
 
 Claude's summary:

 Key optimizations implemented:

  1. Explicit t_powers pre-allocation: # Instead of: torch.stack([torch.ones_like(t), t, t**2, t**3], dim=1) t_powers_buf[:, 0] = 1.0  # Direct assignment, no list creation t_powers_buf[:, 1] = t t_powers_buf[:, 2] = t**2 t_powers_buf[:, 3] = t**3
  2. einops for smart reshaping: # einops creates views when possible, copies only when necessary t_powers = einops.rearrange(t_powers_buf, 'b p -> b 1 1 p') final_control_points[:] = einops.rearrange(width_results, 'p b c -> b c p')
  3. Complete buffer reuse hierarchy:
    - 1D: out + t_powers_buf
    - 2D: width_results + final_control_points + temp_1d + t_powers_buf
    - 3D: depth_results + final_control_points_3d + temp_2d + t_powers_buf_3d + nested buffers_2d
    - 4D: time_results + final_control_points_4d + temp_3d + t_powers_buf_4d + nested buffers_3d

  Performance impact:
  - Memory: Eliminates the exponential allocation pattern (was: 1→5→20→80 tensors, now: fixed buffer reuse)
  - GPU efficiency: Better memory coalescing, reduced fragmentation
  - einops benefits: Smart view creation where possible, minimal copying

  The recursive memory explosion is now contained with proper buffer management, while einops handles the reshaping operations efficiently. This should dramatically
  improve your GPU interpolation performance!

### CODE REVIEW

**Comment by alisterburt** [comment]
this removed a whole bunch of memory wait time but it seems like memory was handled async and I was actually compute bottlenecked...

Key Changes in the Profile
Before:

BufferedReader.read(): 10,353ms (84.1%)
interpolate_grid_1d: 6,980ms (56.7%)
interpolate_pieces_1d: 2,950ms (34.1%)

After:

BufferedReader.read(): Still there but much less dominant
interpolate_grid_1d: 7,221ms (83.4%) - now the clear bottleneck
interpolate_pieces_1d: 2,950ms (34.1%) - similar time

**Comment by McHaillet** [comment]
Seems like you got a nice boost as the 'BufferedReader' looks completely gone :)

Does in-place assigment to the t_powers_buf not break differentiation?

**Comment by McHaillet** [comment]
How big are the tensors you are operating on? If they are small it might also explain why GPU performance is not great

**Comment by alisterburt** [comment]
Thanks for taking a look! Unfortunately the time measurements are misleading, overall runtime was pretty much identical - I'm sampling a few hundred points from a small set of control points about 12k times

Currently toying with a native implementation... need to play a bit more before I decide how to move forward :)

**Comment by mgiammar** [comment]
@alisterburt @McHaillet working on moving PRs/Issues over to the monorepo. Performance improvements could be nice. Would you like me to move this discussion over into the monrepo?

**Comment by alisterburt** [comment]
@mgiammar awesome! not worth it in this case I think - this complicates implementation for ~no difference in perf
